virtool/indexes/data.py

```diff
@@ -0,0 +1,303 @@
+    ) -> Optional[IndexFile]:
+        """
+        Uploads a new index file.
+
+        :param reference_id: the reference ID
+        :param index_id: the index ID
+        :param file_type: the type of the file to upload
+        :param name: the name of the new file
+        :param multipart: the file reader
+        :return: the index file
+        """
+        try:
+            index_file = await create_index_file(self._pg, index_id, file_type, name)
+        except IntegrityError:
+            raise ResourceConflictError()
+
+        upload_id = index_file["id"]
+
+        path = join_index_path(self._config.data_path, reference_id, index_id) / name
+
+        try:
+            size = await naive_writer(await multipart(), path)
+        except asyncio.CancelledError:
+            logger.debug("Index file upload aborted: %s", upload_id)
+            await delete_row(self._pg, upload_id, IndexFile)
+            return None
+
+        index_file = await virtool.uploads.db.finalize(
+            self._pg, size, upload_id, IndexFile
+        )
+
+        return IndexFile(**index_file)
+
+    async def finalize(self, ref_id: str, index_id: str) -> Index:
+        """
+        Finalize an index document.
+
+        :param ref_id: the reference ID
+        :param index_id: the index ID
+        :return: the finalized Index
+        """
+        reference = await self._db.references.find_one(ref_id)
+
+        if reference is None:
+            raise ResourceNotFoundError()
+
+        rows = await get_rows(self._pg, IndexFile, "index", index_id)
+
+        results = {f.name: f.type for f in rows}
+
+        if IndexType.fasta not in results.values():
+            raise ResourceConflictError(
+                "A FASTA file must be uploaded in order to finalize index"
+            )
+
+        if reference["data_type"] == "genome":
+            required_files = [f for f in FILES if f != "reference.json.gz"]
+
+            if missing_files := [f for f in required_files if f not in results]:
+                raise ResourceConflictError(
+                    f"Reference requires that all Bowtie2 index files have been uploaded. "
+                    f"Missing files: {', '.join(missing_files)}"
+                )
+
+        await virtool.indexes.db.finalize(
+            self._db, self._pg, self._config.base_url, ref_id, index_id
+        )
+
+        return await self.get(index_id)
+
+    async def find_changes(
+        self, index_id: str, req_query: MultiDictProxy[str]
+    ) -> HistorySearchResult:
+        """
+        Find the virus changes that are included in a given index build.
+
+        :param index_id: the index ID
+        :param req_query: the request query object
+        :return: the changes
+        """
+        if not await self._db.indexes.count_documents({"_id": index_id}):
+            raise ResourceNotFoundError()
+
+        db_query = {"index.id": index_id}
+
+        term = req_query.get("term")
```

Comment:
Move this to the request handler. It has the responsibility of interpreting and validation request parameters.